### PR TITLE
kakoune-lsp: 16.0.0 -> 17.1.1

### PR DESCRIPTION
**Pull request by deviant**
## Description of changes

The Perl patch no longer applied, so along with creating a new version of it, I took the liberty to clean up the package overall.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Result of `nixpkgs-review pr 302761` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kakoune-lsp (kakounePlugins.kakoune-lsp)</li>
  </ul>
</details>

**Review by alexshpilkin — APPROVED** [review]
Verified basic functionality (`clangd` integration).

**Comment by 2xsaiko** [comment]
This (and also the current version of this package) needs a newer Kakoune version, it broke my configuration:

> Error: 2:9: 'add-highlighter': wrong parameter count (NOTE: lsp-inlay-diagnostics-enable requires Kakoune >= 2024)

**Comment by alexshpilkin** [comment]
@2xsaiko With the version here (17.0.1), your config should actually be fully functional with the exception of inlay diagnostics. The version currently in -unstable (16.0.0), on the other hand, will crash, causing everything after the inlay enable command to not be executed. So if anything this PR is an improvement to the status quo.

Actually working inlay diagnostics in kakoune-lsp ≥ 16 do indeed require a git (unreleased) Kakoune as of now, but in that respect we’re at the mercy of @mawww from what I understand, and not the Nixpkgs maintainerhood.

**Comment by alexshpilkin** [comment]
@2xsaiko Right, inlay stuff in kakoune-lsp ≥ 16 requires git (unreleased) Kakoune; but at least ≥ 17 (this PR, or anything including https://github.com/kakoune-lsp/kakoune-lsp/commit/86ceae0ab5c69dba83c143cd9167f1409f8e0e41) will tell you what’s wrong instead of crashing incomprehensibly.

As far as cutting a release of Kakoune, we’re probably at the mercy of @mawww here as opposed to the Nixpkgs maintainerdom—you don’t usually get unreleased versions in Nixpkgs unless there’s no choice.

**Comment by deviant** [comment]
To anyone who wants to use inlay diagnostics with the current stable version of Kakoune, add the following to the `patches` of `kakoune-unwrapped`:

```
# "Support -after switch for flag-lines highlighter"
# Needed by kakoune-lsp v16.0.0+ for inlay diagnostics.
(fetchpatch2 {
  url = "https://github.com/mawww/kakoune/commit/c124c8f517c921bee2017cdf5081f3e8b98e27b9.patch";
  hash = "sha256-R+B29aKb9WGgEmNmVN0ZL0wPWrBFcTYglVBfhCjKENQ=";
  includes = [ "src/highlighters.cc" ];
})
```

I've been using this for a while with no issues.

**Review by alois31 — APPROVED** [review]
The new version seems to work very well, including code actions when perl is not in PATH (which was previously broken).

**Comment by L-as** [comment]
what's blocking this from being merged?

**Comment by QiBaobin** [comment]
it's 17.1.1 for a while now.

**Comment by deviant** [comment]
I've fixed the conflict. Can someone merge this finally? Once that's done it'd be a simple matter for someone to bump it from there to the latest version.

**Comment by deviant** [comment]
Correction: have now fixed the conflict without using the web UI (which somehow produced an incorrect commit). Should be good to go now.

File: pkgs/by-name/ka/kakoune-lsp/Hardcode-perl.patch
Comment (philiptaron) [inline pkgs/by-name/ka/kakoune-lsp/Hardcode-perl.patch:24]: Preference: I'd prefer if this were `@perl@/bin/perl` so that the `substituteAll` would just be `inherit perl`.

File: pkgs/by-name/ka/kakoune-lsp/Hardcode-perl.patch
Comment (philiptaron) [inline pkgs/by-name/ka/kakoune-lsp/Hardcode-perl.patch:45]: Same here.

File: pkgs/by-name/ka/kakoune-lsp/package.nix
Comment (philiptaron) [inline pkgs/by-name/ka/kakoune-lsp/package.nix:10]:
```suggestion
  version = "17.1.1";
```

File: pkgs/by-name/ka/kakoune-lsp/package.nix
Comment (philiptaron) [inline pkgs/by-name/ka/kakoune-lsp/package.nix:17]: - https://github.com/NixOS/nixpkgs/issues/277994
- https://github.com/NixOS/nixpkgs/issues/310373

```suggestion
  src = fetchFromGitHub {
    owner = "kakoune-lsp";
    repo = "kakoune-lsp";
    rev = "v17.1.1";
    hash = "sha256-XBH2pMDiHJNXrx90Lt0IcsbMFUM+X7GAHgiHpdlIdR4=";
  };
```

File: pkgs/by-name/ka/kakoune-lsp/package.nix
Comment (philiptaron) [inline pkgs/by-name/ka/kakoune-lsp/package.nix:23]: - https://github.com/NixOS/nixpkgs/issues/237216
- https://github.com/NixOS/nixpkgs/pull/301350

```suggestion
    (replaceVars ./Hardcode-perl.patch { inherit perl; })
```

File: pkgs/by-name/ka/kakoune-lsp/package.nix
Comment (philiptaron) [inline pkgs/by-name/ka/kakoune-lsp/package.nix:6]: Please run `nixfmt` on this.

**Review by philiptaron — CHANGES_REQUESTED** [review]

**Comment by philiptaron** [comment]
Thanks for keeping this patch running. I'm back in the saddle and looking to merge once a few things are tidied up.

**Comment by deviant** [comment]
I'm not looking to make further changes on this, sorry. In the several months it took for anyone to actually look at this PR, I've entirely stopped using Nix on any of my machines, so this isn't really a priority for me now.

**Comment by philiptaron** [comment]
> I'm not looking to make further changes on this, sorry. 

Understood. Would you mind if I did so?

**Comment by deviant** [comment]
You're welcome to! :) Please try to keep my authorship metadata intact, depending on how you go about this.

**Comment by philiptaron** [comment]
> You're welcome to! :) Please try to keep my authorship metadata intact, depending on how you go about this.

I'll do my very best. 🙇🏻

**Review by philiptaron — APPROVED** [review]

**Comment by L-as** [comment]
Thanks!

**Comment by L-as** [comment]
Funnily enough this broke basically immediately because of #332957

**Comment by philiptaron** [comment]
- https://github.com/kakoune-lsp/kakoune-lsp/pull/764

I can confirm the following update now makes it build.

**Comment by philiptaron** [comment]
> Funnily enough this broke basically immediately because of #332957

- https://github.com/NixOS/nixpkgs/pull/335266

**Pull request merged**
